### ISSUE TITLE
Enable RMSNorm with fp32 upcast/downcast for gpt-oss, a Q3 target model

### ISSUE THREAD

**Issue by moriohara** [torch-spyre/torch-spyre#3580]
### What feature would you like to be added?

Enable RMSNorm with fp32 upcast/downcast for gpt-oss, which is slightly different from Llama, already supported by PR https://github.com/torch-spyre/torch-spyre/pull/2927 (issue https://github.com/torch-spyre/torch-spyre/issues/2508)
```python
        input_dtype = hidden_states.dtype
        hidden_states = hidden_states.to(torch.float32)
        variance = hidden_states.pow(2).mean(-1, keepdim=True)
        hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
        return (self.weight * hidden_states).to(input_dtype)  # main diff with Llama
```

### Why is this feature needed?

GPT-OSS, one of our Q3 target models, requires this capability.

### Any alternative solution that you may want to share?

Offload to CPU, or use fp16 on device, instead of fp32.

### Anything else we need to know?

- Make sure to promote fp16 tensors to fp32 when one or more other inputs are fp32
- Enhance the logic to test staggered input patterns to support multiple inputs with the same staggered EA in https://github.com/torch-spyre/torch-spyre/blob/f442c9a8dee076a09fb8742e5688e3e140bdee7b/torch_spyre/_inductor/propagate_layouts.py#L877

**Comment by manid2** [comment]
Hi @moriohara @joyalbin I have tested this issue using below test script and it is working as expected,

```python
return (weight * hidden_states).to(input_dtype)
```

`weight` is fp16 with EA `STANDARD` and `hidden_states` is EA `DL16_TO_FP32` but the op `mul` **promotes** weight to float32 so inductor scheduler pass `split_multi_ops` splits it and inserts `to_dtype` for weight to convert to fp32 and its EA will be `DL16_TO_FP32`, now EA are matching so the op is valid and no compilation error.

There is no need to insert any custom op to unstagger hidden_states with EA staggered.

test script

```python
$ cat fp32/issue_3580.py
import torch
import torch_spyre  # noqa: F401


def test_rms_norm_gpt_oss():
    torch.manual_seed(42)

    device = torch.device("spyre")
    dtype = torch.bfloat16

    # Tensor initialization
    hidden_states_cpu = torch.randn(4, 64, dtype=dtype)
    weight_cpu = torch.randn(64, dtype=dtype)
    variance_epsilon = 1e-6

    def fn(hidden_states, weight, variance_epsilon):
        input_dtype = hidden_states.dtype
        hidden_states = hidden_states.to(torch.float32)
        variance = hidden_states.pow(2).mean(-1, keepdim=True)
        hidden_states = hidden_states * torch.rsqrt(variance + variance_epsilon)
        return (weight * hidden_states).to(input_dtype)  # main diff with Llama

    # CPU reference calculation
    expected = fn(hidden_states_cpu, weight_cpu, variance_epsilon)

    # Transfer inputs to Spyre device
    hidden_states_dev = hidden_states_cpu.to(device)
    weight_dev = weight_cpu.to(device)

    # Compile with Inductor backend
    compiled_fn = torch.compile(fn, backend="inductor")
    actual = compiled_fn(hidden_states_dev, weight_dev, variance_epsilon).cpu()

    # Numerical verification
    diff = (actual.float() - expected.float()).abs()
    assert diff.max().item() < 0.5, f"max_abs={diff.max().item()}"
    print(f"[PASS] max_abs diff: {diff.max().item():.6f}")


if __name__ == "__main__":
    test_rms_norm_gpt_oss()

$ python fp32/issue_3580.py
[Fast Path Debug] SUCCESS: 87 patches, 30 eval steps, 5 input symbols, 35 total symbols, output_size=8448
[PASS] max_abs diff: 0.015625
```

logs

```
[DEBUG] [spyre.inductor.split_multi_ops] split_multi_ops: 'buf3' has 2 compute ops: [constant] -> final: add
[INFO] [spyre.inductor.split_multi_ops] split_multi_op: 'buf3' -> 1 intermediate buffers
[DEBUG] [spyre.inductor.split_multi_ops] split_multi_ops: 'buf6' has 2 compute ops: [to_dtype] -> final: mul
[INFO] [spyre.inductor.split_multi_ops] split_multi_op: 'buf6' -> 1 intermediate buffers
[INFO] [spyre.inductor.passes] elapsed     3ms  split_multi_ops
[INFO] [spyre.inductor.propagate_layouts] Multi-arg pointwise (buf5): producing 2 candidate output layouts.
[INFO] [spyre.inductor.propagate_layouts] Multi-arg pointwise (buf6): producing 2 candidate output layouts.
[INFO] [spyre.inductor.passes] elapsed    24ms  propagate_spyre_tensor_layouts
[INFO] [spyre.inductor.passes] elapsed     1ms  validate_ops

[INFO] [spyre.inductor.passes] AFTER PRE-SCHEDULING

op9: ComputedBuffer
  buffer=buf9
  layout=FixedTiledLayout('spyre:0', torch.float32, size=[256], stride=[1], device_layout=SpyreTensorLayout(device_size=[8, 32], stride_map =[32, 1], device_dtype=DataFormats.IEEE_FP32, element_arrangement=ElementArrangement.DL16_TO_FP32))
  Pointwise(
  'spyre',
  torch.float32,
  def inner_fn(index):
      i0 = index
      tmp0 = ops.load(arg1_1, i0)
      tmp1 = ops.to_dtype(tmp0, torch.float32, src_dtype=torch.bfloat16, use_compute_types=True)
      return tmp1
  ,
  ranges=[256],
  origin_node=convert_element_type_default,
  origins=OrderedSet([convert_element_type_default]),
  stack_traces = {,
    File "/home/manid2/dt-inductor/torch-spyre-experiments/fp32/issue_3580.py", line 21, in fn,
      return (weight * hidden_states).to(input_dtype)  # main diff with Llama,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^,
  ,
  }
)

op6: ComputedBuffer
  buffer=buf6
  layout=FixedTiledLayout('spyre:0', torch.float32, size=[4, 256], stride=[256, 1], device_layout=SpyreTensorLayout(device_size=[8, 4, 32], stride_map =[32, 256, 1], device_dtype=DataFormats.IEEE_FP32, element_arrangement=ElementArrangement.DL16_TO_FP32))
  allocation={'lx': 512}
  Pointwise(
  'spyre',
  torch.float32,
  def inner_fn(index):
      i0, i1 = index
      tmp0 = ops.load(buf9, i1)
      tmp1 = ops.load(buf9, i1)
      tmp2 = ops.load(buf5, i1 + 256 * i0)
      tmp3 = tmp1 * tmp2
      return tmp3
  ,
  ranges=[4, 256],
  origin_node=None,
  origins=OrderedSet([mul_1]),
  stack_traces = {,
    File "/home/manid2/dt-inductor/torch-spyre-experiments/fp32/issue_3580.py", line 21, in fn,
      return (weight * hidden_states).to(input_dtype)  # main diff with Llama,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^,
  ,
  }
)

op7: ComputedBuffer
  buffer=buf7
  layout=FixedTiledLayout('spyre:0', torch.bfloat16, size=[4, 256], stride=[256, 1], device_layout=SpyreTensorLayout(device_size=[4, 4, 64], stride_map =[64, 256, 1], device_dtype=DataFormats.SEN169_FP16))
  Pointwise(
  'spyre',
  torch.bfloat16,
  def inner_fn(index):
      i0, i1 = index
      tmp0 = ops.load(buf6, i1 + 256 * i0)
      tmp1 = ops.to_dtype(tmp0, torch.bfloat16, src_dtype=torch.float32, use_compute_types=True)
      return tmp1
  ,
  ranges=[4, 256],
  origin_node=convert_element_type_1,
  origins=OrderedSet([convert_element_type_1]),
  stack_traces = {,
    File "/home/manid2/dt-inductor/torch-spyre-experiments/fp32/issue_3580.py", line 21, in fn,
      return (weight * hidden_states).to(input_dtype)  # main diff with Llama,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^,
  ,
  }
)

[PASS] max_abs diff: 0.031250
```

**Comment by michihirohorie** [comment]
PR #3400 enables type promotion of mul with fp16 and fp32:
```
self.weight * hidden_states
```

**Comment by moriohara** [comment]
@manid2 @michihirohorie 
Thank you so much. I also confirmed at my end. Let's add a test case, perhaps extending the following one, to let CI test this use case to close this issue.

https://github.com/torch-spyre/torch-spyre/blob/9fd980f3ac2c0c6e0df144a033cdca1687e250b6/tests/inductor/test_inductor_normalization_scalars.py#L148-L152

**Comment by moriohara** [comment]
Closing since PR #4044 has added a test case for this use case.
